api/core/v1alpha1/device_types.go

```diff
@@ -122,6 +163,44 @@ type DeviceStatus struct {
 	Conditions []metav1.Condition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type"`
 }
 
+type ProvisioningInfo struct {
+	StartTime  metav1.Time `json:"startTime"`
+	EndTime    metav1.Time `json:"endTime,omitempty"`
+	RebootTime metav1.Time `json:"reboot,omitempty"`
+	Token      string      `json:"token"`
+	Error      string      `json:"error,omitempty"`
+}
+
+func (d *Device) GetActiveProvisioning() *ProvisioningInfo {
+	for i := range d.Status.Provisioning {
+		if d.Status.Provisioning[i].EndTime.IsZero() {
+			return &d.Status.Provisioning[i]
+		}
+	}
+	return nil
+}
+
+func (d *Device) CreateProvisioningEntry() (*ProvisioningInfo, error) {
+	if d.Status.Phase != DevicePhaseProvisioning {
+		return nil, fmt.Errorf("device is in phase %s, expected %s", d.Status.Phase, DevicePhaseProvisioning)
+	}
+	active := d.GetActiveProvisioning()
+	if active != nil {
+		return nil, fmt.Errorf("device has an active provisioning with StartTime %s", active.StartTime.String())
+	}
+	token := make([]byte, 32)
+	_, err := rand.Read(token)
+	if err != nil {
+		return nil, err
+	}
+	entry := ProvisioningInfo{
+		StartTime: metav1.Now(),
+		Token:     hex.EncodeToString(token),
+	}
+	d.Status.Provisioning = append(d.Status.Provisioning, entry)
+	return &d.Status.Provisioning[len(d.Status.Provisioning)-1], nil
```

Comment:
```suggestion
	return &entry, nil
```
As we just appended it, we have the last item.